{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0ae7d86f",
   "metadata": {},
   "source": [
    "# Australian AA Variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed5c4f3",
   "metadata": {},
   "source": [
    "## Part 00: Extract and Transform Metadata\n",
    "\n",
    "The following cells manipulate the GISAID source metadata file,\\\n",
    "extracting sequences collected in Austrailian Regions/Territories.\n",
    "\n",
    "The location of the metadata file (`metadata.tsv.xz`) is specified in\\\n",
    "the cell under \"Definitions > Filepaths\".\n",
    "\n",
    "Several transformed files featuring general summaries will be created\\\n",
    "for use in further analyses.\n",
    "\n",
    "##### General Sequence\n",
    "From the metadata, accession ids are categorized by quarter-year periods (Q1, ...)\n",
    "\n",
    "Using the union of all observed substitutions across all sequences as features, \\\n",
    "amino acid substitutions are coded in a binary manner for each sequence.\n",
    "\n",
    "Then, comparing the count of *all* peer sequences in the region-period,\\\n",
    "will provide the incidence of each substitution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc5e49ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "import helpers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7771aa99",
   "metadata": {},
   "source": [
    "### Definitions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "118530f2",
   "metadata": {},
   "source": [
    "#### Filepaths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d294224f",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = Path('./data')\n",
    "\n",
    "gisaid_metadata_fpath = DATA_DIR / 'metadata_sample.tsv'\n",
    "au_metadata_fpath = DATA_DIR / 'au_metadata_sample.tsv'\n",
    "au_reduced_metadata_fpath = DATA_DIR / 'au_reduced_metadata_sample.tsv'\n",
    "\n",
    "SUMMARY_DIR = DATA_DIR / 'summaries'\n",
    "aa_subs_by_seq_fpath = SUMMARY_DIR / 'aa_subs_by_seq.csv'\n",
    "aa_sub_counts_by_region_period_fname = SUMMARY_DIR / 'aa_sub_counts_by_regional_period.csv'\n",
    "seq_counts_fpath = SUMMARY_DIR / 'seq_counts_per_regional_period.csv'\n",
    "\n",
    "# Uncomment to create the directories\n",
    "# SUMMARY_DIR.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "445e690c",
   "metadata": {},
   "source": [
    "#### DataFrame Column and Index Levels and Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb09fd80",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_levels = [0,1,2,3,4]\n",
    "index_names=['region', 'collection_date', 'virus_code', 'accession_id', 'time_period']\n",
    "short_index_names = ['region', 'time_period']\n",
    "header_levels=[0,1,2,3]\n",
    "header_names = ['gene', 'position', 'aa_ref', 'aa_sub']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e9ec2a9",
   "metadata": {},
   "source": [
    "### Extract, Clean and Condense Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f40b85f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if au_metadata_fpath.exists():  # load cached data\n",
    "    au_md_df = pd.read_csv(\n",
    "        au_metadata_fpath,\n",
    "        sep='\\t',\n",
    "        **helpers.CleanedMetadataImportConfig.import_args\n",
    "    ).fillna(helpers.CleanedMetadataImportConfig.nafills)\n",
    "else:\n",
    "    md_df = pd.read_csv(\n",
    "        gisaid_metadata_fpath,\n",
    "        sep='\\t', \n",
    "        **helpers.RawMetadataImportConfig.import_args\n",
    "    ).fillna(helpers.RawMetadataImportConfig.nafills)\n",
    "\n",
    "    # Filter for human hosts only\n",
    "    md_df = md_df[md_df.Host == 'Human']\n",
    "\n",
    "    md_df = helpers.format_column_names(md_df)\n",
    "\n",
    "    # Locations must include \"Australia\"\n",
    "    au_md_df = md_df[md_df['location'].str.contains(pat='[A|a]ustralia')]\n",
    "    # Get all \"complete\" Australian human host data \n",
    "    au_md_df = au_md_df[\n",
    "        (au_md_df['is_low_coverage'] != True) &\n",
    "        (au_md_df['is_complete']==True)\n",
    "    ]\n",
    "\n",
    "    au_md_df = au_md_df[[\n",
    "        'accession_id',\n",
    "        'virus_name',\n",
    "        'location',\n",
    "        'collection_date',\n",
    "        'submission_date',\n",
    "        'clade',\n",
    "        'variant',\n",
    "        'aa_substitutions',\n",
    "    ]].sort_values('collection_date')\n",
    "\n",
    "    # Cache\n",
    "    au_md_df.to_csv(\n",
    "        au_metadata_fpath,\n",
    "        sep='\\t',\n",
    "        index=False\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "67084e6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>virus_name</th>\n",
       "      <th>location</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>clade</th>\n",
       "      <th>variant</th>\n",
       "      <th>aa_substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>EPI_ISL_9362964</td>\n",
       "      <td>hCoV-19/Australia/NSW-ICPMR-19829/2022</td>\n",
       "      <td>Oceania / Australia / New South Wales / Sydney</td>\n",
       "      <td>2022-01-12</td>\n",
       "      <td>2022-02-01</td>\n",
       "      <td>GRA</td>\n",
       "      <td>VOC Omicron GRA (B.1.1.529+BA.*) first detecte...</td>\n",
       "      <td>(NSP5_P132H,Spike_H69del,Spike_T95I,Spike_A67V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>EPI_ISL_3187212</td>\n",
       "      <td>hCoV-19/Australia/NSW-ICPMR-1316/2021</td>\n",
       "      <td>Oceania / Australia / New South Wales / Sydney</td>\n",
       "      <td>2021-07-27</td>\n",
       "      <td>2021-08-02</td>\n",
       "      <td>GK</td>\n",
       "      <td>VOC Delta GK (B.1.617.2+AY.*) first detected i...</td>\n",
       "      <td>(N_G215C,N_D63G,N_R203M,NSP12_G671S,NSP16_Q238...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>EPI_ISL_7772590</td>\n",
       "      <td>hCoV-19/Australia/VIC28863/2021</td>\n",
       "      <td>Oceania / Australia / Victoria</td>\n",
       "      <td>2021-12-08</td>\n",
       "      <td>2021-12-17</td>\n",
       "      <td>GK</td>\n",
       "      <td>VOC Delta GK (B.1.617.2+AY.*) first detected i...</td>\n",
       "      <td>(N_G215C,NSP4_S34F,Spike_T95I,N_D63G,N_R203M,N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>EPI_ISL_3834119</td>\n",
       "      <td>hCoV-19/Australia/NSW-ICPMR-5506/2021</td>\n",
       "      <td>Oceania / Australia / New South Wales / Sydney</td>\n",
       "      <td>2021-08-16</td>\n",
       "      <td>2021-09-01</td>\n",
       "      <td>GK</td>\n",
       "      <td>VOC Delta GK (B.1.617.2+AY.*) first detected i...</td>\n",
       "      <td>(N_G215C,NS3_A59V,N_D63G,N_R203M,NSP2_V258I,NS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EPI_ISL_520750</td>\n",
       "      <td>hCoV-19/Australia/VIC3037/2020</td>\n",
       "      <td>Oceania / Australia / Victoria</td>\n",
       "      <td>2020-07-03</td>\n",
       "      <td>2020-08-25</td>\n",
       "      <td>GR</td>\n",
       "      <td></td>\n",
       "      <td>(Spike_S477N,N_R203K,N_G204R,NSP2_I120F,NSP12_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession_id                              virus_name  \\\n",
       "546  EPI_ISL_9362964  hCoV-19/Australia/NSW-ICPMR-19829/2022   \n",
       "214  EPI_ISL_3187212   hCoV-19/Australia/NSW-ICPMR-1316/2021   \n",
       "443  EPI_ISL_7772590         hCoV-19/Australia/VIC28863/2021   \n",
       "253  EPI_ISL_3834119   hCoV-19/Australia/NSW-ICPMR-5506/2021   \n",
       "36    EPI_ISL_520750          hCoV-19/Australia/VIC3037/2020   \n",
       "\n",
       "                                           location collection_date  \\\n",
       "546  Oceania / Australia / New South Wales / Sydney      2022-01-12   \n",
       "214  Oceania / Australia / New South Wales / Sydney      2021-07-27   \n",
       "443                  Oceania / Australia / Victoria      2021-12-08   \n",
       "253  Oceania / Australia / New South Wales / Sydney      2021-08-16   \n",
       "36                   Oceania / Australia / Victoria      2020-07-03   \n",
       "\n",
       "    submission_date clade                                            variant  \\\n",
       "546      2022-02-01   GRA  VOC Omicron GRA (B.1.1.529+BA.*) first detecte...   \n",
       "214      2021-08-02    GK  VOC Delta GK (B.1.617.2+AY.*) first detected i...   \n",
       "443      2021-12-17    GK  VOC Delta GK (B.1.617.2+AY.*) first detected i...   \n",
       "253      2021-09-01    GK  VOC Delta GK (B.1.617.2+AY.*) first detected i...   \n",
       "36       2020-08-25    GR                                                      \n",
       "\n",
       "                                      aa_substitutions  \n",
       "546  (NSP5_P132H,Spike_H69del,Spike_T95I,Spike_A67V...  \n",
       "214  (N_G215C,N_D63G,N_R203M,NSP12_G671S,NSP16_Q238...  \n",
       "443  (N_G215C,NSP4_S34F,Spike_T95I,N_D63G,N_R203M,N...  \n",
       "253  (N_G215C,NS3_A59V,N_D63G,N_R203M,NSP2_V258I,NS...  \n",
       "36   (Spike_S477N,N_R203K,N_G204R,NSP2_I120F,NSP12_...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "au_md_df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a027fad7",
   "metadata": {},
   "source": [
    "#### Note on Collection Date vs Submission Date\n",
    "\n",
    "Preferably, all samples would be assigned a quarter based on\\\n",
    "the Collection Date alone.\n",
    "\n",
    "However, some regions have many collection dates which\\\n",
    "are set to January 1st of a given year (e.g., 2020, 2021, 2022)\n",
    "\n",
    "Since it is unlikely that all sequences were collected on Jan. 1st (especially Jan. 1st, 2020),\n",
    "which happens to be the default date-time if only a year is provided on submission,\n",
    "it is assumed that these are information deficient for purposes of defining a specific quarter year.\n",
    "\n",
    "This error was discovered after reviewing a first pass of the data,\n",
    "finding some apparent Jan. 1st collection dates would place variants (such as omicron) \\\n",
    "quarters before initial detection. \n",
    "\n",
    "With the extreme unlikelihood of a previously unknown variant having 10s-to-100s of samples \n",
    "all collected on January 1st without being reported for more than half a year, especially with\\\n",
    "the date also being the default if only a year was provided upon submission, \\\n",
    "it is assumed the submission date was reasonably soon after the actual collection.\n",
    "\n",
    "Given the proportion of samples from certain regions which only provided the year,\n",
    "the Submission Date was instead used as a proxy. This comes with a few limitations and caveats:\n",
    "\n",
    "1. The time between the collection and submission occured across a quarter.\n",
    "2. The time between the collection and submission was exceptionally long, i.e., across more than one quarter\n",
    "\n",
    "However, results after the adjustment generally align with known origin dates of variants,\n",
    "and so this heuristic was sufficient for purposes of this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "20199538",
   "metadata": {},
   "outputs": [],
   "source": [
    "if au_reduced_metadata_fpath.exists():  # load cached data\n",
    "    au_reduced_df = pd.read_csv(\n",
    "        au_reduced_metadata_fpath,\n",
    "        sep='\\t'\n",
    "    )\n",
    "else:\n",
    "    au_reduced_df = au_md_df.copy()\n",
    "\n",
    "    # filter all January 1st dates\n",
    "    jan_first_filter = (\n",
    "        (au_reduced_df.collection_date == pd.to_datetime('2020-01-01')) |\n",
    "        (au_reduced_df.collection_date == pd.to_datetime('2021-01-01')) |\n",
    "        (au_reduced_df.collection_date == pd.to_datetime('2022-01-01'))\n",
    "    )\n",
    "    au_reduced_df.loc[jan_first_filter, 'collection_date'] = au_reduced_df['submission_date']\n",
    "\n",
    "    # Split into year, month, and day columns (day is arbitrary)\n",
    "    au_reduced_df['year'] = au_reduced_df['collection_date'].dt.year\n",
    "    au_reduced_df['month'] = au_reduced_df['collection_date'].dt.month\n",
    "    au_reduced_df['day'] = au_reduced_df['collection_date'].dt.day\n",
    "\n",
    "    # split location string into separate columns\n",
    "    expanded_loc_cols = ['continent', 'country', 'region', 'city']\n",
    "    au_reduced_df[expanded_loc_cols] = au_reduced_df['location'].str.split('/', expand=True)\n",
    "\n",
    "    # Format to succinct \"Virus Code\"\n",
    "    au_reduced_df['virus_code'] = helpers.get_formatted_virus_code(au_reduced_df['virus_name'])\n",
    "\n",
    "    # Create columns for deliniating region-periods\n",
    "    au_reduced_df['quarter'] = au_reduced_df.apply(helpers.get_quarter, axis=1)\n",
    "    au_reduced_df['time_period'] = au_reduced_df.apply(helpers.get_period, axis=1)\n",
    "\n",
    "    au_reduced_df['region'] = au_reduced_df['region'].str.strip()\n",
    "    au_reduced_df['region'] = au_reduced_df['region'].str.replace('?', '', regex=False)\n",
    "\n",
    "    au_reduced_df = au_reduced_df[[\n",
    "        'accession_id',\n",
    "        'virus_name',\n",
    "        'virus_code',\n",
    "        'collection_date',\n",
    "        'submission_date',\n",
    "        'year',\n",
    "        'month',\n",
    "        'day',\n",
    "        'quarter',\n",
    "        'time_period',\n",
    "        # 'continent',  # must be Australia\n",
    "        # 'country',  # must be Australia\n",
    "        'region',\n",
    "        'clade',\n",
    "        'variant',\n",
    "        'aa_substitutions',\n",
    "    ]]\n",
    "\n",
    "    au_reduced_df = au_reduced_df.sort_values('collection_date')\n",
    "\n",
    "    # cache\n",
    "    au_reduced_df.to_csv(\n",
    "        au_reduced_metadata_fpath,\n",
    "        sep='\\t',\n",
    "        index=False\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "291da49b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>virus_name</th>\n",
       "      <th>virus_code</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>day</th>\n",
       "      <th>quarter</th>\n",
       "      <th>time_period</th>\n",
       "      <th>region</th>\n",
       "      <th>clade</th>\n",
       "      <th>variant</th>\n",
       "      <th>aa_substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>EPI_ISL_592143</td>\n",
       "      <td>hCoV-19/Australia/VIC12997/2020</td>\n",
       "      <td>VIC-00012997</td>\n",
       "      <td>2020-08-01</td>\n",
       "      <td>2020-10-22</td>\n",
       "      <td>2020</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Victoria</td>\n",
       "      <td>GR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(Spike_S477N,N_R203K,NSP15_S261P,N_G204R,NSP2_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>EPI_ISL_7987421</td>\n",
       "      <td>hCoV-19/Australia/NSW-ICPMR-17597/2021</td>\n",
       "      <td>NSW-ICPMR-00017597</td>\n",
       "      <td>2021-12-12</td>\n",
       "      <td>2021-12-23</td>\n",
       "      <td>2021</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>GK</td>\n",
       "      <td>VOC Delta GK (B.1.617.2+AY.*) first detected i...</td>\n",
       "      <td>(N_G215C,Spike_T95I,NS3_A59V,N_D63G,N_R203M,NS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>EPI_ISL_663953</td>\n",
       "      <td>hCoV-19/Australia/VIC17624/2020</td>\n",
       "      <td>VIC-00017624</td>\n",
       "      <td>2020-07-25</td>\n",
       "      <td>2020-11-30</td>\n",
       "      <td>2020</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>Victoria</td>\n",
       "      <td>GR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(N_R203K,N_G204R,NSP2_I120F,NSP12_P323L,Spike_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>EPI_ISL_3833749</td>\n",
       "      <td>hCoV-19/Australia/NSW-ICPMR-5136/2021</td>\n",
       "      <td>NSW-ICPMR-00005136</td>\n",
       "      <td>2021-08-20</td>\n",
       "      <td>2021-09-01</td>\n",
       "      <td>2021</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>GK</td>\n",
       "      <td>VOC Delta GK (B.1.617.2+AY.*) first detected i...</td>\n",
       "      <td>(N_G215C,Spike_T95I,N_D63G,N_R203M,NSP12_G671S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>EPI_ISL_456630</td>\n",
       "      <td>hCoV-19/Australia/VIC1731/2020</td>\n",
       "      <td>VIC-00001731</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>2020-06-02</td>\n",
       "      <td>2020</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Victoria</td>\n",
       "      <td>O</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(NSP2_V198I,NSP2_P91S,NS8_S24L,NSP6_A287V,NSP3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession_id                              virus_name  \\\n",
       "109   EPI_ISL_592143         hCoV-19/Australia/VIC12997/2020   \n",
       "454  EPI_ISL_7987421  hCoV-19/Australia/NSW-ICPMR-17597/2021   \n",
       "93    EPI_ISL_663953         hCoV-19/Australia/VIC17624/2020   \n",
       "263  EPI_ISL_3833749   hCoV-19/Australia/NSW-ICPMR-5136/2021   \n",
       "26    EPI_ISL_456630          hCoV-19/Australia/VIC1731/2020   \n",
       "\n",
       "             virus_code collection_date submission_date  year  month  day  \\\n",
       "109        VIC-00012997      2020-08-01      2020-10-22  2020      8    1   \n",
       "454  NSW-ICPMR-00017597      2021-12-12      2021-12-23  2021     12   12   \n",
       "93         VIC-00017624      2020-07-25      2020-11-30  2020      7   25   \n",
       "263  NSW-ICPMR-00005136      2021-08-20      2021-09-01  2021      8   20   \n",
       "26         VIC-00001731      2020-04-29      2020-06-02  2020      4   29   \n",
       "\n",
       "     quarter  time_period           region clade  \\\n",
       "109        3            3         Victoria    GR   \n",
       "454        4            8  New South Wales    GK   \n",
       "93         3            3         Victoria    GR   \n",
       "263        3            7  New South Wales    GK   \n",
       "26         2            2         Victoria     O   \n",
       "\n",
       "                                               variant  \\\n",
       "109                                                NaN   \n",
       "454  VOC Delta GK (B.1.617.2+AY.*) first detected i...   \n",
       "93                                                 NaN   \n",
       "263  VOC Delta GK (B.1.617.2+AY.*) first detected i...   \n",
       "26                                                 NaN   \n",
       "\n",
       "                                      aa_substitutions  \n",
       "109  (Spike_S477N,N_R203K,NSP15_S261P,N_G204R,NSP2_...  \n",
       "454  (N_G215C,Spike_T95I,NS3_A59V,N_D63G,N_R203M,NS...  \n",
       "93   (N_R203K,N_G204R,NSP2_I120F,NSP12_P323L,Spike_...  \n",
       "263  (N_G215C,Spike_T95I,N_D63G,N_R203M,NSP12_G671S...  \n",
       "26   (NSP2_V198I,NSP2_P91S,NS8_S24L,NSP6_A287V,NSP3...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "au_reduced_df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd9d487a",
   "metadata": {},
   "source": [
    "### AA-Substitution Vectorization\n",
    "\n",
    "The following cells build binary vectors for each known substitution."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47124acb",
   "metadata": {},
   "source": [
    "#### Extract the Amino Acid Substitutions (\"AA_Substitutions\" column)\n",
    "\n",
    "Each `aa_substitution` identifier string takes the form:\n",
    "    \n",
    "    {gene}_{reference_aa}{position}{substitute_aa}\n",
    "\n",
    "This is split into the constituent parts to produce a multi-leveled column, ordered as:\n",
    "\n",
    "1. Gene\n",
    "2. Position\n",
    "3. Reference AA\n",
    "4. Substituted AA\n",
    "\n",
    "Note the inversion of the position and reference aa.\n",
    "\n",
    "Then, each row describes the presence (or lack thereof) of the substitution within the sample.\n",
    "\n",
    "For example, `spike_D614G` would be split into a hierarchical column:\n",
    "\n",
    "1. Spike\n",
    "2. 614\n",
    "3. D\n",
    "4. G\n",
    "\n",
    "and would contain a `0` in `EPI_ISL_402125` (the reference sequence), \\\n",
    "but a `1` for delta variant samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "608e014c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 42.4 ms, sys: 294 µs, total: 42.7 ms\n",
      "Wall time: 40.5 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if aa_subs_by_seq_fpath.exists():\n",
    "    aa_subs_by_seq_df = pd.read_csv(\n",
    "        aa_subs_by_seq_fpath,\n",
    "        index_col=index_levels,\n",
    "        header=header_levels\n",
    "        # index_names=index_names\n",
    "    ).fillna(0)\n",
    "else:\n",
    "    # iterate through substitutions, remove parentheses, split by comma. \n",
    "    # If substitution exists, mark with `1`\n",
    "    aa_subs = {\n",
    "        idx: pd.Series(\n",
    "            [1 for _ in row[1:-1].split(',')], index=row[1:-1].split(',')\n",
    "        ) for idx, row in au_reduced_df.set_index(index_names)['aa_substitutions'].items()\n",
    "    }\n",
    "    aa_subs_by_seq_df = pd.DataFrame.from_dict(aa_subs, orient='index')\n",
    "    aa_subs_by_seq_df = aa_subs_by_seq_df.fillna(0).sort_index(axis=1).drop(columns='', errors='ignore')\n",
    "    aa_subs_by_seq_df.index = aa_subs_by_seq_df.index.rename(index_names)\n",
    "    aa_subs_by_seq_df = aa_subs_by_seq_df.astype('int32')\n",
    "\n",
    "    # create levels\n",
    "    aa_multi_cols = aa_subs_by_seq_df.columns.str.extractall(r\"(?P<gene>.+?)_(?P<aa_ref>([A-Z]+?)|(ins))(?P<position>[0-9]+)(?P<aa_sub>.+)\")\n",
    "    aa_multi_cols = aa_multi_cols.fillna('').droplevel(1).drop([2,3], axis=1)\n",
    "    aa_multi_cols = aa_multi_cols.set_index(aa_subs_by_seq_df.columns)\n",
    "    aa_multi_cols = aa_multi_cols[['gene', 'position', 'aa_ref', 'aa_sub']]\n",
    "    \n",
    "    # Apply to new dataframe\n",
    "    aa_subs_by_seq_df.columns = pd.MultiIndex.from_tuples(\n",
    "        aa_multi_cols.values.tolist(), \n",
    "        names=aa_multi_cols.columns.tolist()\n",
    "    )\n",
    "\n",
    "    aa_subs_by_seq_df = aa_subs_by_seq_df.sort_index(axis=1).sort_index(axis=0)\n",
    "\n",
    "    if '.xz' not in str(aa_subs_by_seq_fpath):\n",
    "        # create a fake \"sparse\" table to reduce character count\n",
    "        aa_subs_by_seq_df.replace(0, '').to_csv(aa_subs_by_seq_fpath)\n",
    "    else:\n",
    "        aa_subs_by_seq_df.to_csv(aa_subs_by_seq_fpath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c7cce9f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th colspan=\"5\" halign=\"left\">E</th>\n",
       "      <th colspan=\"5\" halign=\"left\">M</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">Spike</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>14</th>\n",
       "      <th>62</th>\n",
       "      <th>68</th>\n",
       "      <th>74</th>\n",
       "      <th>9</th>\n",
       "      <th>101</th>\n",
       "      <th>162</th>\n",
       "      <th>168</th>\n",
       "      <th>175</th>\n",
       "      <th>19</th>\n",
       "      <th>...</th>\n",
       "      <th>937</th>\n",
       "      <th>95</th>\n",
       "      <th>950</th>\n",
       "      <th>954</th>\n",
       "      <th>964</th>\n",
       "      <th>969</th>\n",
       "      <th>981</th>\n",
       "      <th>982</th>\n",
       "      <th>983</th>\n",
       "      <th>986</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>aa_ref</th>\n",
       "      <th>V</th>\n",
       "      <th>V</th>\n",
       "      <th>S</th>\n",
       "      <th>L</th>\n",
       "      <th>T</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>I</th>\n",
       "      <th>T</th>\n",
       "      <th>Q</th>\n",
       "      <th>...</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>D</th>\n",
       "      <th>Q</th>\n",
       "      <th>K</th>\n",
       "      <th>N</th>\n",
       "      <th>L</th>\n",
       "      <th>S</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>aa_sub</th>\n",
       "      <th>del</th>\n",
       "      <th>F</th>\n",
       "      <th>F</th>\n",
       "      <th>M</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>N</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>L</th>\n",
       "      <th>I</th>\n",
       "      <th>N</th>\n",
       "      <th>H</th>\n",
       "      <th>E</th>\n",
       "      <th>K</th>\n",
       "      <th>F</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>Q</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>virus_code</th>\n",
       "      <th>accession_id</th>\n",
       "      <th>time_period</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Victoria</th>\n",
       "      <th>2020-07-17</th>\n",
       "      <th>VIC-00009767</th>\n",
       "      <th>EPI_ISL_565703</th>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Western Australia</th>\n",
       "      <th>2022-01-21</th>\n",
       "      <th>WA-00001054</th>\n",
       "      <th>EPI_ISL_9163967</th>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">New South Wales</th>\n",
       "      <th>2021-09-12</th>\n",
       "      <th>NSW-ICPMR-00007866</th>\n",
       "      <th>EPI_ISL_4552282</th>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2021-12-04</th>\n",
       "      <th>NSW-ICPMR-00016361</th>\n",
       "      <th>EPI_ISL_7621421</th>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2021-08-09</th>\n",
       "      <th>NSW-ICPMR-00003290</th>\n",
       "      <th>EPI_ISL_3426161</th>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 749 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene                                                                                E  \\\n",
       "position                                                                           14   \n",
       "aa_ref                                                                              V   \n",
       "aa_sub                                                                            del   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                           62   \n",
       "aa_ref                                                                              V   \n",
       "aa_sub                                                                              F   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                           68   \n",
       "aa_ref                                                                              S   \n",
       "aa_sub                                                                              F   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                           74   \n",
       "aa_ref                                                                              L   \n",
       "aa_sub                                                                              M   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                            9   \n",
       "aa_ref                                                                              T   \n",
       "aa_sub                                                                              I   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            1.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                M  \\\n",
       "position                                                                          101   \n",
       "aa_ref                                                                              R   \n",
       "aa_sub                                                                              K   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          162   \n",
       "aa_ref                                                                              K   \n",
       "aa_sub                                                                              N   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          168   \n",
       "aa_ref                                                                              I   \n",
       "aa_sub                                                                              L   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          175   \n",
       "aa_ref                                                                              T   \n",
       "aa_sub                                                                              M   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                           19   \n",
       "aa_ref                                                                              Q   \n",
       "aa_sub                                                                              E   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            1.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                              ...  \\\n",
       "position                                                                          ...   \n",
       "aa_ref                                                                            ...   \n",
       "aa_sub                                                                            ...   \n",
       "region            collection_date virus_code         accession_id    time_period  ...   \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            ...   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            ...   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            ...   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            ...   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            ...   \n",
       "\n",
       "gene                                                                             Spike  \\\n",
       "position                                                                           937   \n",
       "aa_ref                                                                               S   \n",
       "aa_sub                                                                               L   \n",
       "region            collection_date virus_code         accession_id    time_period         \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3             0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9             0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7             0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8             0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7             0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                           95   \n",
       "aa_ref                                                                              T   \n",
       "aa_sub                                                                              I   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            1.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          950   \n",
       "aa_ref                                                                              D   \n",
       "aa_sub                                                                              N   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            1.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            1.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            1.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          954   \n",
       "aa_ref                                                                              Q   \n",
       "aa_sub                                                                              H   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            1.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          964   \n",
       "aa_ref                                                                              K   \n",
       "aa_sub                                                                              E   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          969   \n",
       "aa_ref                                                                              N   \n",
       "aa_sub                                                                              K   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            1.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          981   \n",
       "aa_ref                                                                              L   \n",
       "aa_sub                                                                              F   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          982   \n",
       "aa_ref                                                                              S   \n",
       "aa_sub                                                                              A   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \\\n",
       "position                                                                          983   \n",
       "aa_ref                                                                              R   \n",
       "aa_sub                                                                              C   \n",
       "region            collection_date virus_code         accession_id    time_period        \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0   \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0   \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0   \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0   \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0   \n",
       "\n",
       "gene                                                                                   \n",
       "position                                                                          986  \n",
       "aa_ref                                                                              K  \n",
       "aa_sub                                                                              Q  \n",
       "region            collection_date virus_code         accession_id    time_period       \n",
       "Victoria          2020-07-17      VIC-00009767       EPI_ISL_565703  3            0.0  \n",
       "Western Australia 2022-01-21      WA-00001054        EPI_ISL_9163967 9            0.0  \n",
       "New South Wales   2021-09-12      NSW-ICPMR-00007866 EPI_ISL_4552282 7            0.0  \n",
       "                  2021-12-04      NSW-ICPMR-00016361 EPI_ISL_7621421 8            0.0  \n",
       "                  2021-08-09      NSW-ICPMR-00003290 EPI_ISL_3426161 7            0.0  \n",
       "\n",
       "[5 rows x 749 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_subs_by_seq_df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2047c3ec",
   "metadata": {},
   "source": [
    "## Summary Statistics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45a117b5",
   "metadata": {},
   "source": [
    "### Substitution Count by each Region-Period\n",
    "\n",
    "Calculate the sum of each substitution column, partitioned by combined region-period  (e.g., New Queensland Q1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a8efe0cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_sub_counts_by_region_period = aa_subs_by_seq_df.groupby(level=['region', 'time_period']).sum()\n",
    "aa_sub_counts_by_region_period.to_csv(aa_sub_counts_by_region_period_fname)\n",
    "\n",
    "# fill missing time periods with \"0\"\n",
    "regions = aa_sub_counts_by_region_period.index.get_level_values(0).unique()\n",
    "max_period = aa_sub_counts_by_region_period.index.get_level_values(1).max()\n",
    "\n",
    "aa_sub_counts_by_region_period = aa_sub_counts_by_region_period.reindex([\n",
    "    (region, period) \n",
    "    for region in regions \n",
    "    for period in range(1,max_period + 1)\n",
    "]).fillna(0).astype('int32')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8d73e858",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th colspan=\"5\" halign=\"left\">E</th>\n",
       "      <th colspan=\"5\" halign=\"left\">M</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">Spike</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>14</th>\n",
       "      <th>62</th>\n",
       "      <th>68</th>\n",
       "      <th>74</th>\n",
       "      <th>9</th>\n",
       "      <th>101</th>\n",
       "      <th>162</th>\n",
       "      <th>168</th>\n",
       "      <th>175</th>\n",
       "      <th>19</th>\n",
       "      <th>...</th>\n",
       "      <th>937</th>\n",
       "      <th>95</th>\n",
       "      <th>950</th>\n",
       "      <th>954</th>\n",
       "      <th>964</th>\n",
       "      <th>969</th>\n",
       "      <th>981</th>\n",
       "      <th>982</th>\n",
       "      <th>983</th>\n",
       "      <th>986</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>aa_ref</th>\n",
       "      <th>V</th>\n",
       "      <th>V</th>\n",
       "      <th>S</th>\n",
       "      <th>L</th>\n",
       "      <th>T</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "      <th>I</th>\n",
       "      <th>T</th>\n",
       "      <th>Q</th>\n",
       "      <th>...</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>D</th>\n",
       "      <th>Q</th>\n",
       "      <th>K</th>\n",
       "      <th>N</th>\n",
       "      <th>L</th>\n",
       "      <th>S</th>\n",
       "      <th>R</th>\n",
       "      <th>K</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>aa_sub</th>\n",
       "      <th>del</th>\n",
       "      <th>F</th>\n",
       "      <th>F</th>\n",
       "      <th>M</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>N</th>\n",
       "      <th>L</th>\n",
       "      <th>M</th>\n",
       "      <th>E</th>\n",
       "      <th>...</th>\n",
       "      <th>L</th>\n",
       "      <th>I</th>\n",
       "      <th>N</th>\n",
       "      <th>H</th>\n",
       "      <th>E</th>\n",
       "      <th>K</th>\n",
       "      <th>F</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>Q</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region</th>\n",
       "      <th>time_period</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Queensland</th>\n",
       "      <th>7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tasmania</th>\n",
       "      <th>7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Australian Capital Territory</th>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>South Australia</th>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Western Australia</th>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Northern Territory</th>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>South Australia</th>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Queensland</th>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Australian Capital Territory</th>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tasmania</th>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 749 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene                                       E               M                 \\\n",
       "position                                  14 62 68 74  9 101 162 168 175 19   \n",
       "aa_ref                                     V  V  S  L  T   R   K   I   T  Q   \n",
       "aa_sub                                   del  F  F  M  I   K   N   L   M  E   \n",
       "region                       time_period                                      \n",
       "Queensland                   7             0  0  0  0  0   0   0   0   0  0   \n",
       "Tasmania                     7             0  0  0  0  0   0   0   0   0  0   \n",
       "Australian Capital Territory 6             0  0  0  0  0   0   0   0   0  0   \n",
       "South Australia              8             0  0  0  0  1   0   0   0   0  1   \n",
       "Western Australia            8             0  0  0  0  0   0   0   0   0  0   \n",
       "Northern Territory           6             0  0  0  0  0   0   0   0   0  0   \n",
       "South Australia              3             0  0  0  0  0   0   0   0   0  0   \n",
       "Queensland                   3             0  0  0  0  0   0   0   0   0  0   \n",
       "Australian Capital Territory 4             0  0  0  0  0   0   0   0   0  0   \n",
       "Tasmania                     8             0  0  0  0  0   0   0   0   0  0   \n",
       "\n",
       "gene                                      ... Spike                         \\\n",
       "position                                  ...   937 95 950 954 964 969 981   \n",
       "aa_ref                                    ...     S  T   D   Q   K   N   L   \n",
       "aa_sub                                    ...     L  I   N   H   E   K   F   \n",
       "region                       time_period  ...                                \n",
       "Queensland                   7            ...     0  0   3   0   0   0   0   \n",
       "Tasmania                     7            ...     0  0   0   0   0   0   0   \n",
       "Australian Capital Territory 6            ...     0  0   0   0   0   0   0   \n",
       "South Australia              8            ...     1  3   2   1   0   1   1   \n",
       "Western Australia            8            ...     0  0   0   0   0   0   0   \n",
       "Northern Territory           6            ...     0  0   1   0   0   0   0   \n",
       "South Australia              3            ...     0  0   0   0   0   0   0   \n",
       "Queensland                   3            ...     0  0   0   0   0   0   0   \n",
       "Australian Capital Territory 4            ...     0  0   0   0   0   0   0   \n",
       "Tasmania                     8            ...     0  0   0   0   0   0   0   \n",
       "\n",
       "gene                                                  \n",
       "position                                 982 983 986  \n",
       "aa_ref                                     S   R   K  \n",
       "aa_sub                                     A   C   Q  \n",
       "region                       time_period              \n",
       "Queensland                   7             0   0   0  \n",
       "Tasmania                     7             0   0   0  \n",
       "Australian Capital Territory 6             0   0   0  \n",
       "South Australia              8             0   0   0  \n",
       "Western Australia            8             0   0   0  \n",
       "Northern Territory           6             1   0   0  \n",
       "South Australia              3             0   0   0  \n",
       "Queensland                   3             0   0   0  \n",
       "Australian Capital Territory 4             0   0   0  \n",
       "Tasmania                     8             0   0   0  \n",
       "\n",
       "[10 rows x 749 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_sub_counts_by_region_period.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e69b8731",
   "metadata": {},
   "source": [
    "### Sequence Counts by Regional Period\n",
    "\n",
    "Calculate the number of Australian sequences, partitioned by combined region-period"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cd9e6e34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 9.64 ms, sys: 0 ns, total: 9.64 ms\n",
      "Wall time: 6.49 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if seq_counts_fpath.exists():\n",
    "    ## This will take at least 7 seconds to retrieve\n",
    "    seq_counts_df = pd.read_csv(\n",
    "        seq_counts_fpath,\n",
    "        index_col=short_index_names\n",
    "    )\n",
    "else:\n",
    "    seq_counts_df = aa_subs_by_seq_df.droplevel([1,2]).index.to_frame()\n",
    "    \n",
    "    seq_counts_df = seq_counts_df.reset_index(drop=True)\n",
    "    seq_counts_df = seq_counts_df.groupby(short_index_names).count()\n",
    "    seq_counts_df = seq_counts_df.rename(columns={'accession_id': 'counts'})\n",
    "\n",
    "    seq_counts_df.to_csv(seq_counts_fpath)\n",
    "\n",
    "# fill missing time periods with \"0\"\n",
    "regions = seq_counts_df.index.get_level_values(0).unique()\n",
    "max_period = seq_counts_df.index.get_level_values(1).max()\n",
    "\n",
    "seq_counts_df = seq_counts_df.reindex([\n",
    "    (region, period) \n",
    "    for region in regions \n",
    "    for period in range(1,max_period + 1)\n",
    "]).fillna(0).astype('int32')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "13ad41f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "region                        time_period\n",
       "Western Australia             8               0\n",
       "Tasmania                      2               1\n",
       "Queensland                    3               0\n",
       "Tasmania                      8               0\n",
       "Northern Territory            2               0\n",
       "                              8               1\n",
       "South Australia               9               5\n",
       "Queensland                    6               3\n",
       "Tasmania                      5               0\n",
       "Australian Capital Territory  8              13\n",
       "Name: counts, dtype: int32"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_counts_df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bc689a",
   "metadata": {},
   "source": [
    "## Test\n",
    "\n",
    "Find incidence of Spike D614G across region-periods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "f048cb9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th colspan=\"9\" halign=\"left\">Australian Capital Territory</th>\n",
       "      <th>New South Wales</th>\n",
       "      <th>...</th>\n",
       "      <th>Victoria</th>\n",
       "      <th colspan=\"9\" halign=\"left\">Western Australia</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>time_period</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>1</th>\n",
       "      <th>...</th>\n",
       "      <th>9</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>position</th>\n",
       "      <th>aa_ref</th>\n",
       "      <th>aa_sub</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Spike</th>\n",
       "      <th>614</th>\n",
       "      <th>D</th>\n",
       "      <th>G</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "region                       Australian Capital Territory                      \\\n",
       "time_period                                             1   2   3   4   5   6   \n",
       "gene  position aa_ref aa_sub                                                    \n",
       "Spike 614      D      G                               1.0 NaN NaN NaN NaN NaN   \n",
       "\n",
       "region                                      New South Wales  ... Victoria  \\\n",
       "time_period                     7    8    9               1  ...        9   \n",
       "gene  position aa_ref aa_sub                                 ...            \n",
       "Spike 614      D      G       1.0  1.0  1.0            0.25  ...      1.0   \n",
       "\n",
       "region                       Western Australia                                \\\n",
       "time_period                                  1    2   3   4    5   6   7   8   \n",
       "gene  position aa_ref aa_sub                                                   \n",
       "Spike 614      D      G                    0.5  1.0 NaN NaN  1.0 NaN NaN NaN   \n",
       "\n",
       "region                             \n",
       "time_period                     9  \n",
       "gene  position aa_ref aa_sub       \n",
       "Spike 614      D      G       1.0  \n",
       "\n",
       "[1 rows x 72 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_sub_counts_by_region_period.div(seq_counts_df['counts'].values, axis=0).loc[\n",
    "    :,\n",
    "    aa_sub_counts_by_region_period.columns.get_level_values('position') == '614'\n",
    "].T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907795da",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Australian Seasonal SARS-CoV2 Mutations",
   "language": "python",
   "name": "covmut _au"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
